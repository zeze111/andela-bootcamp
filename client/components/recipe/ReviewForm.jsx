import React, { Component } from 'react';
import PropTypes from 'prop-types';

/**
 *
 *
 * @class ReviewForm
 * @extends {React.Component}
 */
class ReviewForm extends Component {
  /**
   * @description Constructor Function
   * @param {any} props
   * @memberof Home
   * @return {void}
   */
  constructor(props) {
    super(props);
    this.state = {
      title: '',
      comment: '',
    };

    this.onChange = this.onChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }

  /**
   * @param {any} event
   * @memberof Home
   * @return {void}
   */
  onChange(event) {
    this.setState({ [event.target.name]: event.target.value });
  }

  /**
   * @param {any} event
   * @memberof Home
   * @return {void}
   */
  onSubmit(event) {
    event.preventDefault();

    const { recipe, limit, offset } = this.props;
    this.props.reviewRecipe(recipe.id, this.state)
      .then(() => {
        this.props.getReviews(recipe.id, limit, offset);
        this.setState({
          title: '',
          comment: '',
        });
      });
    this.refs.reviewForm.reset();
  }

  /**
   * @param {any} props
   * @memberof Home
   * @return {void}
   */
  render() {
    return (
      <div>
        <h5 className="text1" id="rev"> Reviews </h5>
        <div className="col s12 reviews-style">
          <p className="text2" > Add a comment to review this recipe </p>
          <div className="row no-top">
            <form onSubmit={this.onSubmit} className="col l5 m10 s12" ref="reviewForm"> <br />
              <div >
                <div className="input-field text3">
                  <input
                    id="title"
                    name="title"
                    type="text"
                    placeholder="Title"
                    onChange={this.onChange}
                  />
                </div>
                <div className="input-field text3">
                  <textarea
                    id="cmt"
                    name="comment"
                    className="materialize-textarea "
                    placeholder="Comment"
                    onChange={this.onChange}
                  />
                </div>
                <div className="right-align">
                  <input className="btn grey" type="submit" value="Review" /> <br /> <br />
                </div>
              </div>
            </form>
          </div>
        </div>
      </div>
    );
  }
}

ReviewForm.propTypes = {
  reviewRecipe: PropTypes.func.isRequired,
  getReviews: PropTypes.func.isRequired,
  recipe: PropTypes.objectOf(PropTypes.any).isRequired,
  limit: PropTypes.number.isRequired,
  offset: PropTypes.number.isRequired
};


export default ReviewForm;
