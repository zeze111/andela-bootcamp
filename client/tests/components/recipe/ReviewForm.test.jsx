import React from 'react';

import ReviewForm from '../../../components/recipe/ReviewForm';
import store from '../../../store';

const props = {
  reviewRecipe: jest.fn(() => Promise.resolve()),
  getReviews: jest.fn(() => Promise.resolve()),
  offset: 3,
  limit: 0,
  recipe: {
    id: 1,
    name: 'some thing',
    userId: 12
  }
};

const setup = () => {
  const wrapper = shallow(<ReviewForm {...props} store={store} />);
  return { wrapper }
}

describe('onChange', () => {
  it('should change event', () => {
    const event = {
      target: { name: 'comment', value: 'Love this a lot' },
    };
    const { wrapper } = setup();
    const comment = wrapper.find('#cmt');
    comment.simulate('change', event);
    expect(wrapper.instance().state.comment).toBe('Love this a lot');
  });
});

describe('onSubmit', () => {
  it('be called when form is submitted', () => {
    const { wrapper } = setup();
    const event = {
      preventDefault: jest.fn(),
    };

    document.body.innerHTML =
    '<div>' +
    '  <form id="reviewForm" />' +
    '  <button id="button" />' +
    '</div>';
    wrapper.setState({
      title: 'amazing',
      comment: 'Love this a lot'
    });

    wrapper.instance().onSubmit(event);
    expect(props.reviewRecipe.mock.calls.length).toEqual(1);
  });
});

describe('Test for Review Form', () => {
  it('should render component', () => {
    const wrapper = setup();
    expect(wrapper).toMatchSnapshot()
  });
});
