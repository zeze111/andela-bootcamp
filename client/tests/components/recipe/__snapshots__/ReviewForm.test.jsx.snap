// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for Review Form should render component 1`] = `
Object {
  "wrapper": <div>
    <div
      className="col m9 l8"
    >
      <h5
        className="text1 text-headers"
        id="rev"
      >
         Reviews 
      </h5>
      <div
        className="col s12 reviews-style"
      >
        <p
          className="text2"
        >
           Add a comment to review this recipe 
        </p>
      </div>
    </div>
    <div
      className="row no-top"
    >
      <form
        className="col l5 m7 s10 push-s1 push-m1 push-l1"
        id="reviewForm"
        onSubmit={[Function]}
      >
        <br />
        <div>
          <div
            className="input-field text3"
          >
            <input
              id="title"
              name="title"
              onChange={[Function]}
              placeholder="Title"
              type="text"
            />
          </div>
          <div
            className="input-field text3"
          >
            <textarea
              className="materialize-textarea "
              id="cmt"
              name="comment"
              onChange={[Function]}
              placeholder="Comment"
            />
          </div>
          <div
            className="right-align"
          >
            <input
              className="btn grey"
              type="submit"
              value="Review"
            />
             
            <br />
             
            <br />
          </div>
        </div>
      </form>
    </div>
  </div>,
}
`;
